common/timer.go

```diff
@@ -152,74 +185,114 @@ func (tstore *TimerStore) setNextTimeout(ctx sdk.Context, which types.TimerType,
 	store.Set([]byte(types.NextTimerKey[which]), b)
 }
 
-func (tstore *TimerStore) addTimer(ctx sdk.Context, which types.TimerType, value uint64, data string) {
+func (tstore *TimerStore) addTimer(ctx sdk.Context, which types.TimerType, value uint64, key []byte, data []byte) {
 	store := tstore.getStoreTimer(ctx, which)
-	store.Set(types.EncodeKey(value), []byte(data))
+	timerKey := types.EncodeBlockAndKey(value, key)
+	store.Set(timerKey, data)
 
 	nextValue := tstore.getNextTimeout(ctx, which)
 	if value < nextValue {
 		tstore.setNextTimeout(ctx, which, value)
 	}
 }
 
-func (tstore *TimerStore) delTimer(ctx sdk.Context, which types.TimerType, value uint64) {
+func (tstore *TimerStore) delTimer(ctx sdk.Context, which types.TimerType, value uint64, key []byte) {
 	store := tstore.getStoreTimer(ctx, which)
-	store.Delete(types.EncodeKey(value))
+	timerKey := types.EncodeBlockAndKey(value, key)
+	if !store.Has(timerKey) {
+		panic(fmt.Sprintf("delTimer which %d block %d key %v: no such timer", which, value, key))
+	}
+	store.Delete(timerKey)
 }
 
-func (tstore *TimerStore) AddTimerByBlockHeight(ctx sdk.Context, block uint64, data string) {
-	tstore.addTimer(ctx, types.BlockHeight, block, data)
+// AddTimerByBlockHeight adds a new timer to expire on a given block height.
+// If a timer for that <block, key> tuple exists, it will be overridden.
+func (tstore *TimerStore) AddTimerByBlockHeight(ctx sdk.Context, block uint64, key []byte, data []byte) {
+	if block <= uint64(ctx.BlockHeight()) {
+		panic(fmt.Sprintf("timer expiry block %d smaller than ctx block %d",
```

Comment:
Should we panic? why not returning an error?
I'm asking because in my badges PR, the expiry block I put for the timer is influenced by the badge's epoch field. Thus, the value of the timer's block can be dependent on user input (I don't think we should panic as a result of user input)

Reply:
It doesn't make sense to ask for a timeout in the past - it will never happen. Your code prune this case and not ask to add such timer. (note that the function does not return an error - it shouldn't).

Reply:
yeah, I check it before creating a timer